### PR TITLE
eigs: add error if matrix is too small.

### LINKED ISSUE #195: Issue with eigs?
I am not sure why this happens, but I guess there should be a problem with eigs:

```
In  [2]: a = [1 2;3 4]

Out [2]: 2x2 Array{Int64,2}:
 1  2
 3  4

In  [3]: eig(a)

Out [3]: ([-0.372281,5.37228],
2x2 Array{Float64,2}:
 -0.824565  -0.415974
  0.565767  -0.909377)

In  [4]: eigs(a)
WARNING: nev should be at most 0

Out [4]: ERROR: ArgumentError("nev must be at least one")
while loading In[4], in expression starting on line 1

 in eigs at linalg/arnoldi.jl:23
 in eigs at linalg/arnoldi.jl:5
```

### LINKED ISSUE #10631: Issue with eigs?
I am not sure why this happens, but I guess there should be a problem with eigs:
```
In  [2]: a = [1 2;3 4]

Out [2]: 2x2 Array{Int64,2}:
 1  2
 3  4

In  [3]: eig(a)

Out [3]: ([-0.372281,5.37228],
2x2 Array{Float64,2}:
 -0.824565  -0.415974
  0.565767  -0.909377)

In  [4]: eigs(a)
WARNING: nev should be at most 0

Out [4]: ERROR: ArgumentError("nev must be at least one")
while loading In[4], in expression starting on line 1

 in eigs at linalg/arnoldi.jl:23
 in eigs at linalg/arnoldi.jl:5
```

### PR DESCRIPTION
**Pull request by jiahao**
Fix JuliaLang/LinearAlgebra.jl#195

Also make warnings clearer - we are actually changing user input when nev and
ncv are inappropriate. Also fix typos in warning messages.

### CODE REVIEW

File: base/linalg/arnoldi.jl [hunk 1]
```diff
@@ -16,14 +16,15 @@ function eigs(A, B;
     isgeneral = B !== I
     sym = issym(A) && !iscmplx
     nevmax=sym ? n-1 : n-2
+    nevmax > 0 || throw(ArgumentError("Input matrix A is too small. Use eigfact instead."))
     if nev > nevmax
+        warn("Adjusting nev from $nev to $nevmax")
```
Comment (jakebolewski) [inline base/linalg/arnoldi.jl]: I feel this clamping behavior should throw an `ArgumentErrror` instead

**Comment by StefanKarpinski** [comment]
This strikes me as completely the wrong choice. The interface should be uniform, not the implementation.

**Comment by jiahao** [comment]
I disagree. `eigs` should be considered an expert routine over `eig` and `eigfact`. Anyone who uses ARPACK _must_ know what they are doing if they are to get any meaningful results out of it. All this PR does right now is improve the error message.

**Comment by StefanKarpinski** [comment]
There are two choices:

1. every caller of `eigs` must handle small matrices specially
2. only the implementation of `eigs` must handle small matrices specially.

It just seems crazy to take the first option rather than the second.

**Comment by jiahao** [comment]
Every caller of `eigs` should realize what they are doing makes no sense on a 2x2 matrix. Consider that `eigs` means to solve an eigenproblem _by specifically a Krylov subspace iteration algorithm_. If we really want to honor that meaning, then not only must we provide implementations of the basic Lanczos and Arnoldi factorizations in base (since we won't have access to ARPACK), but also all the associated machinery for spectral transformations such as shift-and-invert, eigenvalue sorting, and convergence testing. The amount of effort needed to do so would be significant, and disproportionate for the marginal use case of 2x2 and 1x1 matrices, whose eigenproblems can be solved by hand and also by other means like `eigfact`.

**Comment by ViralBShah** [comment]
Yeah, implementing all that stuff for eigs would probably be too much of a chore, as much as I would like this to just work.

What does octave do?

**Comment by jiahao** [comment]
I've checked octave's source code and they do call `eig` under the hood for small matrices not handled by ARPACK "for compatibility with Matlab". I still don't think that this is a good idea. The only use cases I can think of for small matrices on `eigs` are:
- users trying to troubleshoot problems with larger matrices
- users who mistyped `eig`, or
- users who want `eigs` to "just work".

I think surreptitiously calling `eig` under the hood would confuse the first group of users. And the last group of users are for the most part misinformed. Every textbook on numerical linear algebra cautions against treating Krylov methods as black boxes, because they can break in extremely subtle ways.

**Comment by StefanKarpinski** [comment]
You know a lot more about this stuff than I do, so carry on.

**Pull request merged**